homeassistant/config_entries.py

```diff
@@ -1169,6 +1174,54 @@ def _data_to_save(self) -> dict[str, list[dict[str, Any]]]:
         """Return data to save."""
         return {"entries": [entry.as_dict() for entry in self._entries.values()]}
 
+    async def async_remove_device(self, entry_id: str, device_id: str) -> bool:
+        """Remove a device."""
+        if (config_entry := self.async_get_entry(entry_id)) is None:
+            raise UnknownEntry
+
+        if not config_entry.supports_remove_device:
+            raise HomeAssistantError("Config entry does not support device removal")
+
+        dev_reg = device_registry.async_get(self.hass)
+        ent_reg = entity_registry.async_get(self.hass)
+
+        if (device_entry := dev_reg.async_get(device_id)) is None:
+            raise HomeAssistantError("Unknown device")
+
+        if entry_id not in device_entry.config_entries:
+            raise HomeAssistantError("Config entry not in device")
+
+        try:
+            integration = await loader.async_get_integration(
+                self.hass, config_entry.domain
+            )
+        except loader.IntegrationNotFound as exc:
+            raise HomeAssistantError from exc
+        try:
+            component = integration.get_component()
+        except ImportError as exc:
+            raise HomeAssistantError from exc
+
+        # Check with the config entry if it can be removed from the device
+        if not await component.async_config_entry_will_remove_from_device(
+            self.hass, config_entry, device_entry
+        ):
+            return False
+
+        # Remove device entities which belong to the config entry
```

Comment:
This sounds like something that we should always do automatically if we remove a config entry from a device ?

Reply:
Here's a related PR:
https://github.com/home-assistant/core/pull/66209

Reply:
> This sounds like something that we should always do automatically if we remove a config entry from a device

So we move the logic to the device registry; should we do that in this PR or in a follow-up?

Reply:
Follow up is fine.

Reply:
We should do that other PR first as this PR relies on it